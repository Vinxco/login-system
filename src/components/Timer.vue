<template>
  <main class="card">
    <h2>Welcome!</h2>
    <h1>The last time you accessed was</h1>
    <section>
      <div class="timer">
        <div>
          <time>{{ currentDays }}</time>
          <span>Days</span>
        </div>
        <div>
          <time>{{ currentHours }}</time>
          <span>Hours</span>
        </div>
        <div>
          <time>{{ currentMinutes }}</time>
          <span>Minutes </span>
        </div>
        <div>
          <time>{{ currentSeconds }}</time>
          <span>Seconds</span>
        </div>
      </div>
    </section>
    <button @click="logout">Logout</button>
  </main>
</template>

<script>
import firebase from "firebase";
import moment from "moment";
export default {
  name: "Timer",
  data: function() {
    return {
      currentSeconds: null,
      currentMinutes: null,
      currentHours: null,
      currentDays: null
    };
  },
  methods: {
    logout() {
      firebase
        .auth()
        .signOut()
        .then(() => {
          this.$router.replace("login");
        });
    },
    updateCurrentTime() {
      const loginDate = firebase.auth().currentUser.metadata.lastSignInTime;
      const login = moment(loginDate);
      const now = moment();
      const time = now.diff(login, "seconds");
      const secondsInAMInute = 60;
      const minutesInAnHour = 60;
      const hoursInADay = 24;
      const secondsInAnHour = 3600;
      const secondsInADay = 86400;
      const seconds = Math.round(time % secondsInAMInute);
      const minutes = Math.floor((time / secondsInAMInute) % minutesInAnHour);
      const hours = Math.floor((time / secondsInAnHour) % hoursInADay);
      const days = Math.floor(time / secondsInADay);
      this.currentSeconds = seconds < 10 ? "0" + seconds : seconds;
      this.currentMinutes = minutes < 10 ? "0" + minutes : minutes;
      this.currentHours = hours < 10 ? "0" + hours : hours;
      this.currentDays = days < 10 ? "0" + days : days;
    }
  },
  created() {
    setInterval(() => this.updateCurrentTime(), 1 * 1000);
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h1 {
  font-size: 0.9rem;
  color: rgb(180, 178, 178);
}
section {
  width: 100%;
  margin: 4rem 0;
}
.timer {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}
.timer div {
  color: rgb(180, 178, 178);
  display: flex;
  flex-direction: column;
}
time {
  font-size: 2rem;
  font-weight: bold;
}
button {
  width: unset;
}
</style>
