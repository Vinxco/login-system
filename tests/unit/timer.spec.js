import { shallowMount } from "@vue/test-utils";
import Timer from "@/components/Timer.vue";


describe('Timer.vue Implementation Test', () => {
  let wrapper = null

  beforeEach(() => {
    wrapper = shallowMount(Timer)
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('initializes with correct elements', () => {

    // check that the heading text is rendered
    expect(wrapper.findAll('h2').length).toEqual(1)
    expect(wrapper.findAll('h2').at(0).text()).toMatch('Welcome!')

    expect(wrapper.findAll('h1').length).toEqual(1)
    expect(wrapper.findAll('h1').at(0).text()).toMatch('The last time you accessed was')

    // check that all the labels are created
    expect(wrapper.findAll('span').length).toEqual(4)
    expect(wrapper.findAll('span').at(0).text()).toMatch('Days')
    expect(wrapper.findAll('span').at(1).text()).toMatch('Hours')
    expect(wrapper.findAll('span').at(2).text()).toMatch('Minutes')
    expect(wrapper.findAll('span').at(3).text()).toMatch('Seconds')

    // check that all the timers are created
    expect(wrapper.findAll('time').length).toEqual(4)
    expect(wrapper.findAll('time').at(0).text()).toMatch('')

    expect(wrapper.vm.currentDays).toBe(null)

    // Check tha the button is created
    expect(wrapper.findAll('button').length).toEqual(1)
    expect(wrapper.findAll('button').at(0).text()).toMatch('Logout')
    
    // match to the snapshot
    expect(wrapper).toMatchSnapshot();
  })

})

