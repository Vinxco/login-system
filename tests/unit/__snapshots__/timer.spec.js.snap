// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer.vue Implementation Test initializes with correct elements 1`] = `
<main class="card">
  <h2>Welcome!</h2>
  <h1>The last time you accessed was</h1>
  <section>
    <div class="timer">
      <div><time></time> <span>Days</span></div>
      <div><time></time> <span>Hours</span></div>
      <div><time></time> <span>Minutes </span></div>
      <div><time></time> <span>Seconds</span></div>
    </div>
  </section> <button>Logout</button>
</main>
`;
